### ISSUE TITLE
Format indentation issue?

### ISSUE THREAD

**Issue by DT-ZhangJia** [Microsoft/vscode-go#899]
VSCode Version: 1.10.2
OS Version: macOS 10.12.3
Extension: Go 0.6.57
Extension add: goreturns/gofmt

Steps to Reproduce:
1. Type code without indentation.
2. Leave the cursor after the code.
3. Format the code (shift+option+f or save when formatOnSave=true).
4. The cursor will select the last element of that line.
5. The element selected will be replaced when press return key to change line, so that the code is broken.

![fmt](https://cloud.githubusercontent.com/assets/25337614/24547699/bfe5c5d4-1643-11e7-8e13-555a021ce532.gif)

Is there any settings to fix that problem? Really annoying.

**Comment by ramya-rao-a** [comment]
@DT-ZhangJia hmm, nice catch.

I can repro it, but am not sure what is causing it.

Meanwhile if anyone can help figuring this one out, that would be very much appreciated.

**Comment by ramya-rao-a** [comment]
Found the root cause!!

The root cause is that `formatOnSave` feature uses the `TextEditorEdit` api to apply the formatting edits on the file.
Edits made this way have this issue of ending up selecting part of the text that the cursor is on.

I have logged an upstream bug for this in VS Code:
https://github.com/Microsoft/vscode/issues/32058

To fix the issue from the Go extension's side, I have updated to use `WorkspaceEdit` api instead.

The fix will be out in the next update.

Thanks for logging the issue
